test/mockInterfaces.go

```diff
@@ -0,0 +1,302 @@
+	return nil, nil
+}
+
+func (h *TestHostService) CleanupDNSRecords(_ context.Context, _ traffic.Interface) error {
+	return nil
+}
+
+func (h *TestHostService) GetManagedHosts(ctx context.Context, traffic traffic.Interface) ([]v1alpha1.ManagedHost, error) {
+	managed := []v1alpha1.ManagedHost{}
+	for _, host := range traffic.GetHosts() {
+		managedZone, subDomain, err := h.GetManagedZoneForHost(ctx, host, traffic)
+		if err != nil {
+			return nil, err
+		}
+		if managedZone == nil {
+			// its ok for no managedzone to be present as this could be a CNAME or externally managed host
+			continue
+		}
+		dnsRecord, err := h.GetDNSRecord(ctx, subDomain, managedZone, traffic)
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return nil, err
+		}
+		managedHost := v1alpha1.ManagedHost{
+			Host:        host,
+			Subdomain:   subDomain,
+			ManagedZone: managedZone,
+			DnsRecord:   dnsRecord,
+		}
+
+		managed = append(managed, managedHost)
+	}
+	return managed, nil
+}
+
+func (h *TestHostService) CreateDNSRecord(_ context.Context, subDomain string, _ *v1alpha1.ManagedZone, _ v1.Object) (*v1alpha1.DNSRecord, error) {
+	if subDomain == FailCreateDNSSubdomain {
+		return nil, fmt.Errorf(FailCreateDNSSubdomain)
+	}
+	record := v1alpha1.DNSRecord{}
+	return &record, nil
+}
+
+func (h *TestHostService) GetDNSRecord(ctx context.Context, subDomain string, managedZone *v1alpha1.ManagedZone, _ v1.Object) (*v1alpha1.DNSRecord, error) {
+	if subDomain == FailFetchDNSsubdinain {
+		return &v1alpha1.DNSRecord{}, fmt.Errorf(FailFetchDNSsubdinain)
+	}
+	record := &v1alpha1.DNSRecord{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      managedZone.Spec.DomainName,
+			Namespace: managedZone.GetNamespace(),
+		},
+	}
+
+	if err := h.controlClient.Get(ctx, client.ObjectKeyFromObject(record), record); err != nil {
+		return nil, err
+	}
+	return record, nil
+}
+
+func (h *TestHostService) GetManagedZoneForHost(ctx context.Context, host string, t traffic.Interface) (*v1alpha1.ManagedZone, string, error) {
+	hostParts := strings.SplitN(host, ".", 2)
+	if len(hostParts) < 2 {
+		return nil, "", fmt.Errorf("unable to parse host : %s on traffic resource : %s", host, t.GetName())
+	}
+	subDomain := hostParts[0]
+	zone := v1alpha1.ManagedZoneList{}
+
+	err := h.controlClient.List(ctx, &zone, client.InNamespace(t.GetNamespace()))
+	if err != nil {
+		return nil, "", err
+	}
+	if len(zone.Items) == 0 {
+		return nil, subDomain, nil
+	}
+	return &zone.Items[0], subDomain, nil
+}
+
+func (h *TestHostService) AddEndpoints(_ context.Context, gateway traffic.Interface, _ *v1alpha1.DNSRecord) error {
+	hosts := gateway.GetHosts()
+	for _, host := range hosts {
+		if host == FailEndpointsHostname {
+			return fmt.Errorf(FailEndpointsHostname)
+		}
+	}
+	return nil
+}
+
+func NewTestHostService(client client.Client) *TestHostService {
+	return &TestHostService{controlClient: client}
+}
```

Comment:
Why is this making calls via the client. You are using this in unit tests right? So it shouldn't be making calls to get k8s resources via that API.

There might be a need to mock this interface, but there shouldn't be this much code in it and the values that are going to be returned for each function should be explicit. Your tests could easy fail because of bugs in your test code.

Reply:
It's being created with a fake client in the unit tests, so can ignore this comment